trecs/models/mf.py

```diff
@@ -0,0 +1,293 @@
+        representations based on the interaction data from the *most recent* call to
+        `run()`.
+
+        >>> mf.run(50) # data from this will not be used in fitting a new MF model
+        >>> mf.run(50) # data from this WILL be used in fitting a new MF model
+        >>> mf.fit_mf()
+
+    """
+
+    def __init__(  # pylint: disable-all
+        self,
+        num_users=None,
+        num_items=None,
+        num_latent_factors=None,
+        user_representation=None,
+        item_representation=None,
+        actual_user_representation=None,
+        actual_item_representation=None,
+        probabilistic_recommendations=False,
+        seed=None,
+        verbose=False,
+        num_items_per_iter=10,
+        model_params=None,
+        **kwargs
+    ):
+        # check inputs
+        num_users, num_items, num_attributes = validate_user_item_inputs(
+            num_users,
+            num_items,
+            user_representation,
+            item_representation,
+            actual_user_representation,
+            actual_item_representation,
+            100,
+            1250,
+            10,
+            num_latent_factors,
+        )
+        if model_params is None:
+            model_params = {}
+        num_features_vals = non_none_values(
+            model_params.get("features"), num_latent_factors, num_attributes
+        )
+        if len(num_features_vals) > 1:
+            raise ValueError("Number of latent factors is not the same across inputs.")
+
+        self.num_latent_factors = num_attributes
+        self.model_params = model_params
+        self.als_model = None  # initialize empty model
+        self.all_interactions = pd.DataFrame(columns=["user", "item"])  # empty interactions matrix
+
+        # generate user and item representations as needed
+        if user_representation is None:
+            user_representation = np.zeros((num_users, num_attributes))
+        if item_representation is None:
+            item_representation = Generator(seed=seed).binomial(
+                n=1, p=0.5, size=(num_attributes, num_items)
+            )
+        # if the actual item representation is not specified, we assume
+        # that the recommender system's beliefs about the item attributes
+        # are the same as the "true" item attributes
+        if actual_item_representation is None:
+            actual_item_representation = np.copy(item_representation)
+
+        measurements = [MSEMeasurement()]
+
+        super().__init__(
+            user_representation,
+            item_representation,
+            actual_user_representation,
+            actual_item_representation,
+            num_users,
+            num_items,
+            num_items_per_iter,
+            probabilistic_recommendations=probabilistic_recommendations,
+            measurements=measurements,
+            verbose=verbose,
+            seed=seed,
+            **kwargs
+        )
+
+    def _update_internal_state(self, interactions):
```

Comment:
So, in the case that we are doing startup, this makes sense to me. I am not following though, what happens after new interactions are observed, but we are not retraining the model. Is that because we were still waiting to make a determination on that? Or maybe I am just missing something in the code?

Reply:
`_update_internal_state` is run at every single timestep and records the interactions at that timestep. 

`startup_and_train()` is the method that runs multiple startup iterations _and then_ fits an MF model to all of the interactions gathered during those startup iterations. So at the end of `startup_and_train`, there is a call to `update_predicted_scores`, which calls `fit_mf()`. `fit_mf()` is therefore called _after_ some number of iterations, rather than at every single timestep.

The reason this is confusing is because this functionality is only clear if you also look at `recommender.py` to see how the inheritance works in terms of function calls. Sorry about that :/